pyfr/backends/cuda/cublaslt.py

```diff
@@ -0,0 +1,285 @@
+            alpha_ct, beta_ct = c_float(alpha), c_float(beta)
+            ctype = w.COMPUTE_32F
+            dtype = w.R_32F
+
+        ws_ptr = c_void_p()
+        desc = GEMMDesc(None, None, None, None, None, 0)
+
+        # Convenience wrapper
+        def gemm(stream):
+            w.cublasLtMatmul(
+                h, desc.mm, byref(alpha_ct), A, desc.a, B,
+                desc.b, byref(beta_ct), C, desc.c, C, desc.c,
+                byref(desc.algo), ws_ptr, desc.ws_size, stream
+            )
+
+        # Obtain the performance of the kernel
+        try:
+            desc, dt = self._mul_cache[ckey]
+        except KeyError:
+            # Create matrix layouts
+            a_desc = CUBLASLtMatrixLayout(self, A, dtype)
+            b_desc = CUBLASLtMatrixLayout(self, B, dtype)
+            c_desc = CUBLASLtMatrixLayout(self, C, dtype)
+
+            # Matmul descriptor
+            mm_attr = [(w.MATMUL_DESC_TRANSA, w.OP_N, c_int),
+                       (w.MATMUL_DESC_TRANSB, w.OP_N, c_int),
+                       (w.MATMUL_DESC_TRANSC, w.OP_N, c_int)]
+            mm_desc = CUBLASLtMatmulDesc(self, ctype, dtype, mm_attr)
+
+            # Heuristic preference descriptor
+            pref_attr = [(w.MATMUL_PREF_MAX_WORKSPACE_BYTES,
+                          w.WORKSPACE_MAX_SIZE,
+                          c_size_t)]
+            pref = CUBLASLtMatmulPreference(self, pref_attr)
+
+            # Get matmul heuristics
+            heurs = (CUBLASLtMatmulHeuristicResult * self.nkerns)()
+            nreturn = c_int()
+            w.cublasLtMatmulAlgoGetHeuristic(
+                h, mm_desc, a_desc, b_desc, c_desc, c_desc, pref, self.nkerns,
+                heurs, byref(nreturn)
+            )
+
+            # Save a copy of the contents of the output matrix
+            out_np = getattr(out, 'parent', out).get()
+
+            # Allocate some temporary worspace for benchmarks
+            ws_ptr = CUDADevAlloc(cuda, w.WORKSPACE_MAX_SIZE)
+
+            # Benchmark the kernel
+            best_kern = None
+            for heur in heurs[:nreturn.value]:
+                # Allocate gemm workspace
+                ws_size = heur.workspaceSize
+
+                desc = GEMMDesc(
+                    mm_desc, a_desc, b_desc, c_desc, heur.algo, ws_size
+                )
+
+                dt = self._benchmark(gemm)
+                if best_kern is None or dt < best_kern[-1]:
+                    best_kern = desc, dt
+
+            # Restore the output matrix
+            getattr(out, 'parent', out).set(out_np)
+
+            # Update the cache
+            desc, dt = best_kern
+            self._mul_cache[ckey] = best_kern
+
+        class MulKernel(CUDAKernel):
+            def __init__(self, mats, dt, ws_ptr):
+                super().__init__(mats=mats, dt=dt)
+                self.ws_ptr = ws_ptr
+
+            def add_to_graph(self, graph, deps):
+                stream = cuda.create_stream()
+                ws_ptr = self.ws_ptr
```

Comment:
Not sure this will work as you expect (ws_ptr is local and so will not be seen by the GEMM further up).  Either want the obscure `nonlocal` keyword, or just shift the allocation scope.

Reply:
Where should the allocation go? I really wasn't sure about this bit.

Reply:
Just above the class definition.